r/src/array_from_vector.cpp

```diff
@@ -201,6 +202,67 @@ struct VectorToArrayConverter {
     return Status::OK();
   }
 
+  template <typename T>
+  arrow::enable_if_t<is_struct_type<T>::value, Status> Visit(const T& type) {
+    using BuilderType = typename TypeTraits<T>::BuilderType;
+    ARROW_RETURN_IF(!Rf_inherits(x, "data.frame"),
+                    Status::RError("Expecting a data frame"));
+
+    auto* struct_builder = checked_cast<BuilderType*>(builder);
+
+    int64_t n = vctrs::short_vec_size(x);
```

Comment:
Why do you need this?

Reply:
It seems in R it's possible to have data-frame-like tibble columns that are malformed viewed as Arrow-like (as evidenced by the unit test)

Reply:
`vctrs::short_vec_size()` knows how to get the size of things like data frames with 0 columns, data frames with data frame columns ...

Reply:
See also https://jira.apache.org/jira/browse/ARROW-9028 and some skipped tests about tables with 0 columns